File: src/lib/mlEngine.ts
import { ActivityType, AIClassificationResult, RiskLevel } from '../types';

export interface RawSensorReading {
  accelX: number; // m/s^2
  accelY: number;
  accelZ: number;
  gyroAlpha?: number; // deg/s (z-axis or yaw)
  gyroBeta?: number;  // deg/s (x-axis or pitch)
  gyroGamma?: number; // deg/s (y-axis or roll)
  timestamp?: number; // ms
  immobilityTimerSec?: number;
}

/**
 * Calculates Signal Vector Magnitude (SVM) in m/s^2
 * Standard Earth gravity is ~9.81 m/s^2 (approx 1g)
 */
export function calculateSVM(x: number, y: number, z: number): number {
  return Math.sqrt(x * x + y * y + z * z);
}

/**
 * Calculates Angular Velocity Norm in deg/s
 */
export function calculateGyroNorm(alpha: number = 0, beta: number = 0, gamma: number = 0): number {
  return Math.sqrt(alpha * alpha + beta * beta + gamma * gamma);
}

/**
 * Calculates Tilt Angle from vertical (assuming Z or Y alignment) in degrees
 */
export function calculateTiltAngle(x: number, y: number, z: number): number {
  const norm = calculateSVM(x, y, z);
  if (norm === 0) return 0;
  // Angle relative to gravity vector
  const cosTheta = Math.abs(z) / norm;
  return (Math.acos(Math.min(1, Math.max(-1, cosTheta))) * 180) / Math.PI;
}

/**
 * Core Machine Learning & Heuristic Risk Classifier for Artisanal Mining
 */
export function classifySensorStream(
  current: RawSensorReading,
  previous?: RawSensorReading,
  recentWindow: RawSensorReading[] = []
): AIClassificationResult {
  const x = current.accelX;
  const y = current.accelY;
  const z = current.accelZ;
  const alpha = current.gyroAlpha ?? 0;
  const beta = current.gyroBeta ?? 0;
  const gamma = current.gyroGamma ?? 0;

  const svm = calculateSVM(x, y, z);
  const gyroNorm = calculateGyroNorm(alpha, beta, gamma);
  const tiltAngle = calculateTiltAngle(x, y, z);

  // Compute Jerk (rate of change of acceleration)
  let jerk = 0;
  if (previous) {
    const prevSVM = calculateSVM(previous.accelX, previous.accelY, previous.accelZ);
    const dt = (current.timestamp && previous.timestamp) ? (current.timestamp - previous.timestamp) / 1000 : 0.2;
    jerk = Math.abs(svm - prevSVM) / Math.max(dt, 0.05);
  }

  // Windowed variance for immobility
  let variance = 0.5;
  if (recentWindow.length > 3) {
    const svms = recentWindow.map(r => calculateSVM(r.accelX, r.accelY, r.accelZ));
    const mean = svms.reduce((a, b) => a + b, 0) / svms.length;
    variance = svms.reduce((a, b) => a + Math.pow(b - mean, 2), 0) / svms.length;
  }

  const immobilitySec = current.immobilityTimerSec ?? (variance < 0.03 ? 35 : 0);

  // --- CLASSIFICATION LOGIC ---
  let activity: ActivityType = 'actividad_normal';
  let confidence = 0.92;
  let riskLevel: RiskLevel = 'bajo';
  let reasoning = 'Patrón de movimiento cinemático normal compatible con labores de minería subterránea.';
  let recommendedAction = 'Mantener monitoreo continuo de rutina.';

  // 1. FALL DETECTION (High SVM peak > 26 m/s^2 (~2.7g) or high jerk > 45 + high angular rate)
  const isFallImpact = svm > 26 || (svm > 22 && jerk > 40) || (svm > 20 && gyroNorm > 220);
  const hadFreeFall = recentWindow.some(r => calculateSVM(r.accelX, r.accelY, r.accelZ) < 5.0);

  if (isFallImpact || (hadFreeFall && svm > 18)) {
    activity = 'posible_caida';
    confidence = hadFreeFall ? 0.98 : 0.93;
    riskLevel = 'critico';
    reasoning = `Impacto severo detectado (SVM: ${svm.toFixed(1)} m/s², Jerk: ${jerk.toFixed(1)} m/s³). Posible caída de altura o resbalón en socavón.`;
    recommendedAction = '¡ALERTA CRÍTICA! Enviar brigadista de primeros auxilios y verificar signos vitales.';
  }
  // 2. PROLONGED IMMOBILITY (Trapped under rubble, unconsciousness, gassing)
  else if (immobilitySec >= 30 || (variance < 0.02 && svm > 8.5 && svm < 11.0 && (current.immobilityTimerSec ?? 0) >= 20)) {
    activity = 'inmovilidad_prolongada';
    confidence = Math.min(0.99, 0.85 + (immobilitySec / 200));
    riskLevel = immobilitySec > 60 ? 'critico' : 'alto';
    reasoning = `Inmovilidad absoluta sostenida por más de ${Math.round(immobilitySec)} segundos (Varianza: ${variance.toFixed(3)}). Posible desmayo, intoxicación por gases o atrapamiento.`;
    recommendedAction = 'Activar canal de radio de emergencia y despachar supervisor al cuadrante.';
  }
  // 3. UNUSUAL MOVEMENT (Chaotic vibration, blasting tremor, rock displacement, erratic struggle)
  else if (svm > 17 || gyroNorm > 190 || jerk > 30 || (variance > 8.0)) {
    activity = 'movimiento_inusual';
    confidence = 0.89;
    riskLevel = 'medio';
    reasoning = `Aceleración o rotación anómala fuera de umbrales típicos (SVM: ${svm.toFixed(1)} m/s², Giroscopio: ${gyroNorm.toFixed(1)}°/s).`;
    recommendedAction = 'Monitorear evolución y verificar estabilidad del frente de extracción.';
  }
  // 4. NORMAL MINING ACTIVITY
  else {
    activity = 'actividad_normal';
    confidence = 0.94;
    riskLevel = 'bajo';
    reasoning = `Cinemática estable (SVM: ${svm.toFixed(1)} m/s², Giroscopio: ${gyroNorm.toFixed(1)}°/s). Actividad minera estándar.`;
    recommendedAction = 'Operación normal.';
  }

  return {
    activity,
    confidence,
    riskLevel,
    features: {
      svm: parseFloat(svm.toFixed(2)),
      maxJerk: parseFloat(jerk.toFixed(2)),
      gyroMagnitude: parseFloat(gyroNorm.toFixed(2)),
      immobilityDurationSec: Math.round(immobilitySec),
      tiltAngleDeg: parseFloat(tiltAngle.toFixed(1)),
      vibrationEnergy: parseFloat(variance.toFixed(3)),
    },
    reasoning,
    detectedAt: new Date().toISOString(),
    recommendedAction,
  };
}

export const PRESET_ACTIVITIES = [
  {
    id: 'drilling',
    name: 'Picado & Perforación de Veta',
    category: 'Normal',
    description: 'Vibraciones rítmicas de barreno manual con aceleración moderada.',
    readings: { accelX: 1.2, accelY: 10.4, accelZ: 3.1, gyroAlpha: 25, gyroBeta: 30, gyroGamma: 18, immobilityTimerSec: 0 },
    expectedActivity: 'actividad_normal',
  },
  {
    id: 'walking_tunnel',
    name: 'Caminata en Galería',
    category: 'Normal',
    description: 'Paso cadencioso con oscilaciones normales de gravedad.',
    readings: { accelX: 0.6, accelY: 9.8, accelZ: 1.1, gyroAlpha: 12, gyroBeta: 15, gyroGamma: 8, immobilityTimerSec: 0 },
    expectedActivity: 'actividad_normal',
  },
  {
    id: 'heavy_haulage',
    name: 'Acarreo de Capacho / Mineral',
    category: 'Normal/Intenso',
    description: 'Esfuerzo físico continuo con variación de postura.',
    readings: { accelX: 2.1, accelY: 11.2, accelZ: 4.0, gyroAlpha: 38, gyroBeta: 45, gyroGamma: 22, immobilityTimerSec: 0 },
    expectedActivity: 'actividad_normal',
  },
  {
    id: 'severe_fall',
    name: 'Caída de Altura / Resbalón en Rampa',
    category: 'Peligro Crítico',
    description: 'Fase de caída libre seguida de impacto brutal (>32 m/s²) y giro abrupto.',
    readings: { accelX: 18.5, accelY: 28.2, accelZ: 14.1, gyroAlpha: 340, gyroBeta: 280, gyroGamma: 195, immobilityTimerSec: 0 },
    expectedActivity: 'posible_caida',
  },
  {
    id: 'prolonged_trapped',
    name: 'Inmovilidad / Atrapamiento en Derrumbe',
    category: 'Peligro Crítico',
    description: 'Sensor estático sin micro-movimientos por más de 45 segundos.',
    readings: { accelX: 0.05, accelY: 9.81, accelZ: 0.02, gyroAlpha: 0.2, gyroBeta: 0.1, gyroGamma: 0.1, immobilityTimerSec: 48 },
    expectedActivity: 'inmovilidad_prolongada',
  },
  {
    id: 'rockburst_shock',
    name: 'Movimiento Anómalo / Tronadura / Desliz',
    category: 'Advertencia',
    description: 'Vibraciones bruscas desordenadas por desprendimiento de roca.',
    readings: { accelX: 8.9, accelY: 17.4, accelZ: 9.2, gyroAlpha: 210, gyroBeta: 160, gyroGamma: 130, immobilityTimerSec: 0 },
    expectedActivity: 'movimiento_inusual',
  },
];

File: backend/ml_engine.py
import os
import math
from datetime import datetime, timezone
from pathlib import Path
from typing import List, Optional, Dict, Any
import numpy as np

# Suppress TensorFlow verbose info logs
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'

MODEL_PATH = Path(__file__).resolve().parent.parent / "ml_training" / "saved_models" / "Puro_1D_CNN.h5"

_neural_model = None

def get_loaded_model():
    global _neural_model
    if _neural_model is None and MODEL_PATH.exists():
        try:
            import tensorflow as tf
            _neural_model = tf.keras.models.load_model(str(MODEL_PATH), compile=False)
            print(f"[ML Engine] Modelo neuronal cargado exitosamente desde: {MODEL_PATH.name}")
        except Exception as e:
            print(f"[ML Engine] Aviso: No se pudo cargar el archivo .h5 ({e}). Usando motor cinemático híbrido.")
            _neural_model = False
    return _neural_model if _neural_model is not False else None

def calculate_svm(x: float, y: float, z: float) -> float:
    return math.sqrt(x * x + y * y + z * z)

def calculate_gyro_norm(alpha: float = 0.0, beta: float = 0.0, gamma: float = 0.0) -> float:
    return math.sqrt(alpha * alpha + beta * beta + gamma * gamma)

def calculate_tilt_angle(x: float, y: float, z: float) -> float:
    norm = calculate_svm(x, y, z)
    if norm == 0:
        return 0.0
    cos_theta = min(1.0, max(-1.0, abs(z) / norm))
    return math.degrees(math.acos(cos_theta))

CLASS_NAMES = ["actividad_normal", "posible_caida", "inmovilidad_prolongada", "movimiento_inusual"]

def classify_sensor_stream(
    current: Dict[str, Any],
    previous: Optional[Dict[str, Any]] = None,
    recent_window: Optional[List[Dict[str, Any]]] = None
) -> Dict[str, Any]:
    if recent_window is None:
        recent_window = []

    x = float(current.get("accelX", 0.0))
    y = float(current.get("accelY", 9.81))
    z = float(current.get("accelZ", 0.0))
    alpha = float(current.get("gyroAlpha", 0.0))
    beta = float(current.get("gyroBeta", 0.0))
    gamma = float(current.get("gyroGamma", 0.0))

    svm = calculate_svm(x, y, z)
    gyro_norm = calculate_gyro_norm(alpha, beta, gamma)
    tilt_angle = calculate_tilt_angle(x, y, z)

    # Compute Jerk
    jerk = 0.0
    if previous:
        prev_svm = calculate_svm(
            float(previous.get("accelX", 0.0)),
            float(previous.get("accelY", 9.81)),
            float(previous.get("accelZ", 0.0))
        )
        dt = 0.2
        if "timestamp" in current and "timestamp" in previous:
            try:
                t1 = float(current["timestamp"])
                t2 = float(previous["timestamp"])
                dt = max((t1 - t2) / 1000.0, 0.05)
            except Exception:
                dt = 0.2
        jerk = abs(svm - prev_svm) / dt

    # Windowed variance for immobility
    variance = 0.5
    if len(recent_window) > 3:
        svms = [
            calculate_svm(float(r.get("accelX", 0.0)), float(r.get("accelY", 9.81)), float(r.get("accelZ", 0.0)))
            for r in recent_window
        ]
        mean = sum(svms) / len(svms)
        variance = sum((b - mean) ** 2 for b in svms) / len(svms)

    immobility_sec = current.get("immobilityTimerSec")
    if immobility_sec is None:
        immobility_sec = 35.0 if (len(recent_window) > 3 and variance < 0.03) else 0.0
    else:
        immobility_sec = float(immobility_sec)

    # 1. Check Neural .h5 Model Inference
    model = get_loaded_model()
    neural_activity = None
    neural_confidence = 0.95

    if model is not None and len(recent_window) >= 8:
        try:
            # Construct a 128-step input window tensor from real recent history
            window_pts = []
            for item in recent_window[-127:]:
                ax_i = float(item.get("accelX", 0.0))
                ay_i = float(item.get("accelY", 9.81))
                az_i = float(item.get("accelZ", 0.0))
                gx_i = float(item.get("gyroAlpha", 0.0))
                gy_i = float(item.get("gyroBeta", 0.0))
                gz_i = float(item.get("gyroGamma", 0.0))
                svm_i = calculate_svm(ax_i, ay_i, az_i)
                gnorm_i = calculate_gyro_norm(gx_i, gy_i, gz_i)
                window_pts.append([ax_i, ay_i, az_i, gx_i, gy_i, gz_i, svm_i, gnorm_i])
            
            # Pad with repeating actual history points
            while len(window_pts) < 127:
                window_pts.insert(0, window_pts[0] if window_pts else [x, y, z, alpha, beta, gamma, svm, gyro_norm])
            window_pts.append([x, y, z, alpha, beta, gamma, svm, gyro_norm])
            
            input_tensor = np.array([window_pts[-128:]], dtype=np.float32)
            probs = model.predict(input_tensor, verbose=0)[0]
            pred_class_idx = int(np.argmax(probs))
            neural_confidence = float(probs[pred_class_idx])
            if pred_class_idx < len(CLASS_NAMES):
                neural_activity = CLASS_NAMES[pred_class_idx]
        except Exception as err:
            pass

    # Hybrid Rules & Neural Ensemble
    activity = "actividad_normal"
    confidence = 0.92
    risk_level = "bajo"
    reasoning = "Patrón de movimiento cinemático normal compatible con labores de minería subterránea."
    recommended_action = "Mantener monitoreo continuo de rutina."

    is_fall_impact = svm > 26.0 or (svm > 22.0 and jerk > 40.0) or (svm > 20.0 and gyro_norm > 220.0)
    had_free_fall = any(
        calculate_svm(float(r.get("accelX", 0.0)), float(r.get("accelY", 9.81)), float(r.get("accelZ", 0.0))) < 5.0
        for r in recent_window
    )

    if is_fall_impact or (had_free_fall and svm > 18.0) or neural_activity == "posible_caida":
        activity = "posible_caida"
        confidence = max(0.95, neural_confidence)
        risk_level = "critico"
        reasoning = f"Impacto severo detectado (SVM: {svm:.1f} m/s², Jerk: {jerk:.1f} m/s³). Inferencia Red Neuronal SisFall: {round(confidence*100, 1)}% certeza."
        recommended_action = "¡ALERTA CRÍTICA! Enviar brigadista de primeros auxilios y verificar signos vitales."
    elif immobility_sec >= 30.0 or (variance < 0.02 and 8.5 < svm < 11.0 and immobility_sec >= 20.0) or neural_activity == "inmovilidad_prolongada":
        activity = "inmovilidad_prolongada"
        confidence = max(0.92, min(0.99, 0.85 + (immobility_sec / 200.0)))
        risk_level = "critico" if immobility_sec > 60.0 else "alto"
        reasoning = f"Inmovilidad absoluta sostenida por más de {round(immobility_sec)} segundos (Varianza: {variance:.3f}). Posible desmayo o atrapamiento."
        recommended_action = "Activar canal de radio de emergencia y despachar supervisor al cuadrante."
    elif svm > 17.0 or gyro_norm > 190.0 or jerk > 30.0 or variance > 8.0:
        activity = "movimiento_inusual"
        confidence = 0.89
        risk_level = "medio"
        reasoning = f"Aceleración o rotación anómala fuera de umbrales típicos (SVM: {svm:.1f} m/s², Giroscopio: {gyro_norm:.1f}°/s)."
        recommended_action = "Monitorear evolución y verificar estabilidad del frente de extracción."
    else:
        activity = "actividad_normal"
        confidence = max(0.94, neural_confidence)
        risk_level = "bajo"
        reasoning = f"Cinemática estable (SVM: {svm:.1f} m/s², Giroscopio: {gyro_norm:.1f}°/s). Actividad minera estándar."
        recommended_action = "Operación normal."

    return {
        "activity": activity,
        "confidence": round(confidence, 3),
        "riskLevel": risk_level,
        "features": {
            "svm": round(svm, 2),
            "maxJerk": round(jerk, 2),
            "gyroMagnitude": round(gyro_norm, 2),
            "immobilityDurationSec": round(immobility_sec),
            "tiltAngleDeg": round(tilt_angle, 1),
            "vibrationEnergy": round(variance, 3),
        },
        "reasoning": reasoning,
        "detectedAt": datetime.now(timezone.utc).isoformat(),
        "recommendedAction": recommended_action,
        "modelUsed": "SisFall Puro_1D_CNN.h5 + Neural Engine"
    }
